### PR TITLE
Add `state` parameter to Alternatives module

### PR DESCRIPTION
**Pull request by leynos**
##### SUMMARY
Added a `state` parameter to the alternatives module to allow the removal of an alternatives symlink by passing `state=absent`.

In the process of adding this parameter, I have extracted functions where applicable and performed minor tidy-ups to the module per recommendations from pylint.

This tidy-up will enable further improvements I hope to make to this module including the addition of support for dependent alternatives symlinks.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
alternatives

##### ANSIBLE VERSION
```
ansible) [leynos@gradius src]$ ansible --version
ansible 2.4.0
  config file = 
  configured module search path = [u'/home/leynos/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/leynos/Projects/ansible/src/lib/ansible
  executable location = /home/leynos/Projects/ansible/bin/ansible
  python version = 2.7.13 (default, Jan 13 2017, 10:15:16) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

##### ADDITIONAL INFORMATION

In the example below, we will create an alternative `list-files` set to `/usr/bin/ls`, and then remove it.

```
# ./hacking/test-module -m lib/ansible/modules/system/alternatives.py -a "name=list-files path=/usr/bin/ls link=/usr/bin/list-files"
* including generated source, if any, saving to: /root/.ansible_module_generated
* ansiballz module detected; extracted module source to: /root/debug_dir
***********************************
RAW OUTPUT

{"invocation": {"module_args": {"priority": 50, "path": "/usr/bin/ls", "state": "present", "link": "/usr/bin/list-files", "name": "list-files"}}, "changed": true}


***********************************
PARSED OUTPUT
{
    "changed": true, 
    "invocation": {
        "module_args": {
            "link": "/usr/bin/list-files", 
            "name": "list-files", 
            "path": "/usr/bin/ls", 
            "priority": 50, 
            "state": "present"
        }
    }
}
# list-files 
CHANGELOG.md	      ROADMAP.rst		lib
CODING_GUIDELINES.md  VERSION			packaging
CONTRIBUTING.md       ansible-core-sitemap.xml	requirements.txt
COPYING		      bin			setup.py
MANIFEST.in	      contrib			shippable.yml
MODULE_GUIDELINES.md  docs			test
Makefile	      docsite_requirements.txt	ticket_stubs
README.md	      examples			tox.ini
RELEASES.txt	      hacking
# ./hacking/test-module -m lib/ansible/modules/system/alternatives.py -a "name=list-files path=/usr/bin/ls state=absent"
* including generated source, if any, saving to: /root/.ansible_module_generated
* ansiballz module detected; extracted module source to: /root/debug_dir
***********************************
RAW OUTPUT

{"invocation": {"module_args": {"priority": 50, "path": "/usr/bin/ls", "state": "absent", "link": null, "name": "list-files"}}, "changed": true}


***********************************
PARSED OUTPUT
{
    "changed": true, 
    "invocation": {
        "module_args": {
            "link": null, 
            "name": "list-files", 
            "path": "/usr/bin/ls", 
            "priority": 50, 
            "state": "absent"
        }
    }
}
# list-files
bash: /usr/bin/list-files: No such file or directory
```

### CODE REVIEW

**Comment by leynos** [comment]
ready_for_review

File: lib/ansible/modules/system/alternatives.py [hunk 1]
```diff
@@ -132,40 +130,93 @@ def main():
...
     else:
         module.exit_json(changed=False)
 
+def remove_alternative(module, cmd, name, path, all_alternatives):
+    """Remove the requested path if necessary"""
+    if path in all_alternatives:
+        try:
+            module.run_command([cmd, '--remove', name, path], check_rc=True)
+            module.exit_json(changed=True)
+        except subprocess.CalledProcessError:
```
Comment (alikins) [inline lib/ansible/modules/system/alternatives.py]: module.run_command is catching subprocess.CalledProcessError and calling module.fail_json, so should never get this exception at this point.

(looks like that was already in the existing code and not introduced in this pr, so just a note...)

**Comment by alikins** [comment]
Looks good to me, but I haven't tested it.

File: lib/ansible/modules/system/alternatives.py (hunk 1 above)
Comment (leynos) [reply lib/ansible/modules/system/alternatives.py]: @alikins should I remove the try/except blocks around the calls to module.run_command()?

**Comment by leynos** [comment]
Check-mode support was missing from state=absent.  Commit 498426b adds this.

**Comment by aklemp** [comment]
Please also take a look at the proposed extension of the state parameter in #24351.

**Comment by leynos** [comment]
That sounds like a good idea. I'll have a go at implementing that.

**Comment by pilou-** [comment]
@leynos just for you to know, Ansible 2.4 feature freeze for community module is planned for 29 August 2017.

**Comment by leynos** [comment]
@pilou- my apologies. I haven't had any free time at home to work on programming projects. I do intend to return to this.

**Pull request closed without merging**